File: src/lowlevel/server/icefsoperators/icefsGetAttr.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-19 15:29:43
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 14:31:12
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsGetAttr.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"

	"golang.org/x/sys/unix"
)

func (s *IcefsServer) doIcefsGetAttr(fakeInode uint64, statStructBuilder StatStructBuilder) (status int32, attr any) {
	var err error
	var unixStat unix.Stat_t

	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	inode.inodeLock.Lock()
	s.inodeCacheLock.RUnlock()
	err = unix.Fstatat(inode.fd, "", &unixStat, unix.AT_EMPTY_PATH|unix.AT_SYMLINK_NOFOLLOW)
	if err != nil {
		inode.inodeLock.Unlock()
		status = icefserror.IcefsStdErrno(err)
		return
	}
	UnixStatFillSyscallStat(&inode.stat, &unixStat)

	attr = statStructBuilder(&inode.stat)
	inode.inodeLock.Unlock()
	status = icefserror.ICEFS_EOK
	return
}

func (s *IcefsGRpcServer) DoIcefsGetAttr(ctx context.Context, req *pb.IcefsGetAttrReq) (*pb.IcefsGetAttrRes, error) {
	var res pb.IcefsGetAttrRes
	var attr any

	res.Status, attr = s.server.doIcefsGetAttr(req.Inode, GRpcStatStructBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Stat = attr.(*pb.StatStruct)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsGetAttr(ctx context.Context, req *icefsthrift.IcefsGetAttrReq) (*icefsthrift.IcefsGetAttrRes, error) {
	var res icefsthrift.IcefsGetAttrRes
	var attr any

	res.Status, attr = s.server.doIcefsGetAttr(uint64(req.Inode), ThriftStatStructBuilder)
	if res.Status == icefserror.ICEFS_EOK {
		res.Stat = attr.(*icefsthrift.StatStruct)
	}

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsGetAttr.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:47
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsGetAttr.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsGetAttr(fuse_req_t fuseReq, fuse_ino_t inode,
                                 struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsGetAttrReq req;
      icefsgrpc::IcefsGetAttrRes res;
      grpc::ClientContext ctx;
      struct stat attr;

      req.set_inode(inode);

      grpc::Status status = gRpcClient->DoIcefsGetAttr(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        IcefsGRpcFillAttrIn(&attr, res.stat());
        fuse_reply_attr(fuseReq, &attr, this->clientConfig.cacheTimeout);
      } else {
        ICEFS_PR_ERR_STATUS;
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsGetAttrReq req;
      icefsthrift::IcefsGetAttrRes res;
      struct stat attr;

      req.__set_inode(inode);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsGetAttr(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        IcefsThriftFillAttrIn(&attr, res.stat);
        fuse_reply_attr(fuseReq, &attr, this->clientConfig.cacheTimeout);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}
